### PR TITLE
Empty values in QueryParameterTransform and headers *Remove transforms

### LINKED ISSUE #619: QueryParameterTransform should allow URL parameter with value that is empty string
### Describe the bug
I'm proxying request from `http://source/path` to `http://destination/path?param1=&param2=value`, with two transforms:
```c#
new QueryParameterFromStaticTransform(QueryStringTransformMode.Set,"param1","")
new QueryParameterFromStaticTransform(QueryStringTransformMode.Set,"param2","value")
```
However, `param1` is dropped while being applied to `RequestParametersTransformContext` due to [this line of code](https://github.com/microsoft/reverse-proxy/blob/master/src/ReverseProxy/Service/RuntimeModel/Transforms/QueryParameterTransform.cs#L30).

Empty string is valid variable value as per [RFC 6570](https://tools.ietf.org/html/rfc6570) and in my case the destination server cannot work without `param1`.

### To Reproduce

```c#
var context = new RequestParametersTransformContext
{
    Query = new QueryTransformContext(HttpContext.Request)
};
new QueryParameterFromStaticTransform(QueryStringTransformMode.Set, "Empty", "").Apply(context);

// url should equal to ?Empty= but is empty string atm
var url = context.Query.QueryString.ToString();
```

### Further technical details

- Include the version of the packages you are using
**Package version:** 1.0.0-preview.7.20562.2
**ASP.NET version:** .NET Core 3.1 and .NET 5
- The platform (Linux/macOS/Windows)
Windows

### PR DESCRIPTION
**Pull request by alnikola**
Adds support for empty string value in QueryParameterTransform. It changes the way the empty string "" value is handled, now it doesn't remove the given query parameter or header anymore, instead the "" is treated as a normal string value. To remove a query parameter the existing QueryParameterRemoveTransform can be used. To remove headers, the new transforms RequestHeaderRemoveTransform, ResponseHeaderRemoveTransform, and ResponseTrailerRemoveTransforms are added by this PR.
Note, setting "" as a header value is not recommended and can cause an undefined behavior.

Fixes #619

### CODE REVIEW

**Comment by Tratcher** [comment]
Offline notes:
- Can you represent a null value in config?
- Set (empty) is how we've represented removal for Query and Headers. Maybe we should change to an explicit Remove gesture instead.

**Comment by alnikola** [comment]
@Tratcher @MihaZupan The PR is ready for review.

File: src/ReverseProxy/Service/RuntimeModel/Transforms/QueryTransformContext.cs [hunk 1]
```diff
@@ -35,13 +36,8 @@ public QueryString QueryString
...
                 }
 
-#if NET
-                var queryBuilder = new QueryBuilder(_modifiedQueryParameters);
-#elif NETCOREAPP3_1
-                var queryBuilder = new QueryBuilder(_modifiedQueryParameters.SelectMany(kvp => kvp.Value, (kvp, v) => KeyValuePair.Create(kvp.Key, v)));
-#else
-#error A target framework was added to the project and needs to be added to this condition.
-#endif
+                var queryBuilder = new QueryBuilder(_modifiedQueryParameters
```
[the remark this answers is not in the document]
Comment (alnikola) [reply src/ReverseProxy/Service/RuntimeModel/Transforms/QueryTransformContext.cs:39]: No, it's added to support StringValues.Empty case.

[the remark this answers is not in the document]
Comment (alnikola) [reply src/ReverseProxy/Service/RuntimeModel/Transforms/QueryTransformContext.cs:39]: `StringValues.IsNullOrEmpty(kvp.Value)` checks if kvp.Value is empty.

File: src/ReverseProxy/Service/RuntimeModel/Transforms/ResponseHeaderRemoveTransform.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            if (Always || Success(context))
+            {
+                if (!context.ProxyResponse.Headers.Remove(HeaderName))
+                {
+                    context.ProxyResponse.Content?.Headers.Remove(HeaderName);
+                }
```
[the remark this answers is not in the document]
Comment (alnikola) [reply src/ReverseProxy/Service/RuntimeModel/Transforms/ResponseHeaderRemoveTransform.cs]: Fixed

File: src/ReverseProxy/Service/RuntimeModel/Transforms/ResponseTrailerRemoveTransform.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+                var responseTrailersFeature = context.HttpContext.Features.Get<IHttpResponseTrailersFeature>();
+                var responseTrailers = responseTrailersFeature.Trailers;
+                // Support should have already been checked by the caller.
+                Debug.Assert(responseTrailers != null);
+                Debug.Assert(!responseTrailers.IsReadOnly);
+
+                if (!responseTrailers.Remove(HeaderName) && !context.HeadersCopied)
+                {
+                    context.ProxyResponse.TrailingHeaders.Remove(HeaderName);
+                }
```
[the remark this answers is not in the document]
Comment (alnikola) [reply src/ReverseProxy/Service/RuntimeModel/Transforms/ResponseTrailerRemoveTransform.cs]: Fixed

File: test/ReverseProxy.Tests/Service/RuntimeModel/Transforms/ResponseHeaderRemoveTransformTests.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+            var transform = new ResponseHeaderRemoveTransform(removedHeader, always);
+            await transform.ApplyAsync(new ResponseTransformContext()
+            {
+                HttpContext = httpContext,
+                ProxyResponse = proxyResponse,
+                HeadersCopied = true,
+            });
+
+            var expectedHeaders = expected.Split("; ", StringSplitOptions.RemoveEmptyEntries);
+            Assert.Equal(expectedHeaders, proxyResponse.Headers.Select(h => h.Key));
```
[the remark this answers is not in the document]
Comment (alnikola) [reply test/ReverseProxy.Tests/Service/RuntimeModel/Transforms/ResponseHeaderRemoveTransformTests.cs]: Fixed

File: test/ReverseProxy.Tests/Service/RuntimeModel/Transforms/ResponseTrailerRemoveTransformTests.cs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        [InlineData("header1; header2; header3", "value1, value2, value3", 404, false, "headerX", "header1; header2; header3")]
+        [InlineData("header1; header2; header3", "value1, value2, value3", 200, true, "headerX", "header1; header2; header3")]
+        [InlineData("header1; header2; header3", "value1, value2, value3", 404, true, "headerX", "header1; header2; header3")]
+        [InlineData("header1; header2; header2; header3", "value1, value2-1, value2-2, value3", 200, false, "header2", "header1; header3")]
+        [InlineData("header1; header2; header2; header3", "value1, value2-1, value2-2, value3", 404, false, "header2", "header1; header2; header3")]
+        [InlineData("header1; header2; header2; header3", "value1, value2-1, value2-2, value3", 200, true, "header2", "header1; header3")]
+        [InlineData("header1; header2; header2; header3", "value1, value2-1, value2-2, value3", 404, true, "header2", "header1; header3")]
+        public async Task RemoveTrailer_Success(string names, string values, int status, bool always, string removedHeader, string expected)
```
[the remark this answers is not in the document]
Comment (alnikola) [reply test/ReverseProxy.Tests/Service/RuntimeModel/Transforms/ResponseTrailerRemoveTransformTests.cs]: Removed.

File: src/ReverseProxy/Service/RuntimeModel/Transforms/QueryTransformContext.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Tratcher) [reply src/ReverseProxy/Service/RuntimeModel/Transforms/QueryTransformContext.cs:39]: StringValues.Empty is `new string[0];` which is not the same as `new[] { string.Empty };`. StringValues.IsNullOrEmpty(kvp.Value) conflates the two. For query parameters I think we should only support `new[] { string.Empty };`, meaning one query parameter with an empty value. That should work as is with the other constructor, correct?

It's also worth noting that StringValues.Empty is not covered in your tests, and there's no way to express it from the config or existing transforms. The only way to get that value would be with a custom transform.

**Review by Tratcher — APPROVED** [review]

File: src/ReverseProxy/Service/RuntimeModel/Transforms/QueryTransformContext.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (alnikola) [reply src/ReverseProxy/Service/RuntimeModel/Transforms/QueryTransformContext.cs:39]: Empty string values for query parameters are covered in [these several tests](https://github.com/microsoft/reverse-proxy/blob/f98df11067985e93bf86227dd862477d7d7eb7a6/test/ReverseProxy.Tests/Service/RuntimeModel/Transforms/QueryParameterFromStaticTransformTests.cs#L71) added my this PR.

[the remark this answers is not in the document]
Comment (alnikola) [reply src/ReverseProxy/Service/RuntimeModel/Transforms/QueryTransformContext.cs]: I reverted this file.

File: src/ReverseProxy/Service/RuntimeModel/Transforms/QueryTransformContext.cs [hunk 6]
```diff
@@ -15,6 +15,7 @@ namespace Yarp.ReverseProxy.Service.RuntimeModel.Transforms
     /// </summary>
     public class QueryTransformContext
     {
+        private static readonly IEnumerable<string> _emptyString = new[] { string.Empty };
```
Comment (Tratcher) [inline src/ReverseProxy/Service/RuntimeModel/Transforms/QueryTransformContext.cs:18]: No longer used.

Comment (alnikola) [reply src/ReverseProxy/Service/RuntimeModel/Transforms/QueryTransformContext.cs]: Removed

**Pull request merged**